zkevm-circuits/src/evm_circuit/execution/extcodehash.rs

```diff
@@ -0,0 +1,148 @@
+            AccountFieldTag::CodeHash,
+            external_code_hash.expr(),
+        );
+
+        let opcode = cb.query_cell();
+        let step_state_transition = StepStateTransition {
+            rw_counter: Delta(5.expr()),
+            program_counter: Delta(1.expr()),
+            stack_pointer: Delta(0.expr()),
+            ..Default::default()
+        };
+        let dynamic_gas_cost = (1.expr() - previously_accessed.expr())
+            * (GasCost::COLD_ACCOUNT_ACCESS_COST.as_u64()
+                - GasCost::WARM_STORAGE_READ_COST.as_u64())
+            .expr();
+        let same_context =
+            SameContextGadget::construct(cb, opcode, step_state_transition, Some(dynamic_gas_cost));
+
+        Self {
+            same_context,
+            external_address,
+            tx_id,
+            previously_accessed,
+            external_code_hash,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        block: &Block<F>,
+        tx: &Transaction,
+        _: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let external_address = block.rws[step.rw_indices[0]].stack_value().to_address();
+        let mut le_bytes = external_address.0;
+        le_bytes.reverse();
+        self.external_address
+            .assign(region, offset, Some(le_bytes))?;
+
+        self.tx_id
+            .assign(region, offset, U256::from(tx.id).to_scalar())?;
+
+        let previously_accessed = match GasCost::from(step.gas_cost) {
+            GasCost::COLD_ACCOUNT_ACCESS_COST => 0,
+            GasCost::WARM_STORAGE_READ_COST => 1,
+            _ => unreachable!(),
+        };
+        self.previously_accessed
+            .assign(region, offset, Some(F::from(previously_accessed)))?;
+
+        let external_code_hash = block.rws[step.rw_indices[4]].stack_value();
+        self.external_code_hash.assign(
+            region,
+            offset,
+            Some(Word::random_linear_combine(
+                external_code_hash.to_le_bytes(),
+                block.randomness,
+            )),
+        )?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::test_util::run_test_circuits;
+    use eth_types::{address, bytecode, ToWord, H160};
+
+    fn test_ok(address: H160) {
+        let code = bytecode! {
+            PUSH20(address.to_word())
+            #[start]
+            EXTCODEHASH
+            STOP
+        };
+        assert_eq!(run_test_circuits(code), Ok(()));
+    }
+
+    #[test]
+    fn extcodehash_gadget_warm_account() {
+        // The default address is already in the address access set because
+        // `run_test_circuits` works by executing bytecode deployed at the
+        // default address.
+        test_ok(H160::default());
+    }
+
+    #[test]
+    fn extcodehash_gadget_cold_account() {
```

Comment:
I think we should also have a failing test case where the gas available is only sufficient for a warm account, whereas `gas_cost` should be more than that (for a cold account). In this scenario the prover should fail with insufficient gas available.

Reply:
I wrote this test:
```
    #[test]
    fn extcodehash_gadget_insufficient_gas() {
        let address = address!("0xaabbccddee000000000000000000000000000011");
        let code = bytecode! {
            PUSH20(address.to_word())
            #[start]
            EXTCODEHASH
            STOP
        };
        assert_eq!(test_circuits_using_bytecode(code, BytecodeTestConfig {
            gas_limit: 100,
            ..Default::default()
        }), Ok(()));
    }
 ```
but it fails with 
```
---- evm_circuit::execution::extcodehash::test::extcodehash_gadget_insufficient_gas stdout ----
thread 'evm_circuit::execution::extcodehash::test::extcodehash_gadget_insufficient_gas' panicked at 'called `Result::unwrap()` on an `Err` value: TracingError("Failed to trace tx, err: intrinsic gas too low: have 100, want 21000")', zkevm-circuits/src/test_util.rs:58:78
```
which is during the witness generation. I don't think it's possible (right now) to even try to generate a proof for a transaction that reverts because of insufficient gas. 

The two existing warm/cold tests already ensure that the gas cost constraint in the state transition is correct.

Reply:
We're working on enabling better test-case generation. See: #349 
Other Opcodes like `GASPRICE` also do have issues to be implemented until we close it.
We can leave an issue open as tech-debt or something similar.

For now, we don't really have a way as when Geth gets an out-of-gas exception it doesn't return the trace. 
We also need to update this behaviour. And we should file an issue for that.

Reply:
Forget about what I said above. 
I was wrong. In fact, a tx like the one you purpose here is not valid. And would not be included in a block as basically, there's no way to execute it. 

The minimum gas needed to start a tx is 21000. Hence, if you use 21001 as gas argument this would end up in an out-of-gas and you'd get a trace back. But without enough funds to even kickstart the tx, 

My suggestion is to increase the gas you provide to the tx.